### PR TITLE
Don't crash in OOP with TS projects/documents

### PR DESCRIPTION
**Pull request by Pilchie**
## Ask Mode
**Customer scenario**

Enable the option to perform analysis in an external process, and then either:
1. search in solution explorer with a JS or TS file open in the editor.
2. Put the caret on a C# identifier in a cshtml file that also contains a JS script block.

**Bugs this fixes:** https://devdiv.visualstudio.com/DevDiv/_workitems/edit/454792 and https://devdiv.visualstudio.com/DevDiv/_workitems/edit/455588
**Workarounds, if any:** Disable the option to run these features out of process, but we'd really like to get feedback on the option to enable it by default later.
**Risk:** Low - just some checks specifically to prevent these cases, in the OOP specific codepaths.
**Performance impact:** Low - just some if checks.
**Is this a regression from a previous update?** Yes, if the OOP option is enabled.
**Root cause analysis:** Once again, we need to do more testing with JS/TS and razor scenarios.
**How was the bug found?** Divisional automated tests immediately after enabling the option for internal users, and internal dogfooders.

### CODE REVIEW

**Comment by Pilchie** [comment]
I recommend reviewing with ?w=1 for this one.

File: src/Workspaces/Remote/ServiceHub/Services/CodeAnalysisService_DocumentHighlights.cs [hunk 1]
```diff
@@ -16,7 +16,8 @@ internal partial class CodeAnalysisService : IRemoteDocumentHighlights
         {
             var solution = await GetSolutionAsync().ConfigureAwait(false);
             var document = solution.GetDocument(documentId);
-            var documentsToSearch = ImmutableHashSet.CreateRange(documentIdsToSearch.Select(solution.GetDocument));
+            var documentsToSearch = ImmutableHashSet.CreateRange(
```
Comment (heejaechang) [inline src/Workspaces/Remote/ServiceHub/Services/CodeAnalysisService_DocumentHighlights.cs]: removing this to let things crash might make bug more explicit. with this, we might not crash and didn't know that some document just silent failed to run HR. same for FAR.

**Comment by heejaechang** [comment]
👍

File: src/Workspaces/Remote/ServiceHub/Services/CodeAnalysisService_DocumentHighlights.cs (hunk 1 above)
Comment (Pilchie) [inline src/Workspaces/Remote/ServiceHub/Services/CodeAnalysisService_DocumentHighlights.cs]: Without this we *definitely* crash as soon as you edit a .cshtml file with js/ts code in it :smile:

See my question in email about how to filter for highlight references.

**Comment by Pilchie** [comment]
Actually - questions from people posted publicly here:

1.	Highlight references – should we filter on the VS side, the remote side, do the whole search on the VS side, do the non-remotable stuff on the VS side and then merge with the remotable stuff from the OOP side if there is any?
2.	Things that operate on the solution.  We might get partial results, because only the projects that are supported OOP will return results.  It looks like there is at least one “FindDeclarations” API that will fall into this bucket.

**Comment by heejaechang** [comment]
All depends.

Not all language support OOP. So if a feature needs to run on 1 specfic language, then doing filtering on client side seems right.

If a feature needs to run on whole solution regardless of language, then the feature should run both and merge them.

-	heejae

File: src/Features/Core/Portable/AddImport/AbstractAddImportFeatureService.cs [hunk 2]
```diff
@@ -67,7 +67,7 @@ protected AbstractAddImportFeatureService()
 
             using (session)
             {
-                if (session == null)
+                if (session == null || !RemoteSupportedLanguages.IsSupported(document.Project.Language))
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/AddImport/AbstractAddImportFeatureService.cs]: How would this happen?  AbstractAddImportFeatureService is only derived with C# and VB.

File: src/Features/Core/Portable/NavigateTo/AbstractNavigateToSearchService.Remote.cs [hunk 3]
```diff
@@ -31,6 +31,11 @@ internal abstract partial class AbstractNavigateToSearchService
 
         private static Task<RemoteHostClient.Session> GetRemoteHostSessionAsync(Project project, CancellationToken cancellationToken)
         {
+            if (!RemoteSupportedLanguages.IsSupported(project.Language))
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/NavigateTo/AbstractNavigateToSearchService.Remote.cs]: How does this happen.  AbstractNavitateToSearchService is only instantiated for C# and VB.

**Comment by CyrusNajmabadi** [comment]
I don't understand why this change is needed at all.  Several of these services are C#/VB only.  And so they should only be getting C#/VB documents.  As such, i don't see why they'd have a problem with remoting things.

**Comment by Pilchie** [comment]
@CyrusNajmabadi It's quite possible not all of these are needed - I did a bunch of this via code inspection.

The two known instances so far are:

1. A crash in highlight references in a cshtml file that contains both TS and C# - `documentsToSearch` ends up with `null`s in it because it can't re-hydrate the IDs for the TS documents on the OOP side.
2. A crash in solution explorer search in a solution that contains TS file - the `projectId` to search doesn't exist in the remote workspace.

Basically, I think we need to carefully audit, and hopefully provide some safety guarantees for the future about how to guard features that work OOP for C# and VB when they get mixed with languages that don't support OOP.

I'm definitely open to suggestions for better ways to do that.

**Comment by CyrusNajmabadi** [comment]
Interesting.  My primary concern here is that we are actually breaking functionality.  i mean, take: 

> A crash in solution explorer search in a solution that contains TS file - the projectId to search doesn't exist in the remote workspace.

How was this working before?  

--

My ideal would be that the remote workspace has all the project/documents/sourcetexts.  It might be that it doesn't have all the services.  But we would have had to be resilient to that before.  Else, we would have crashed.  And if the services are local, and not on the remote, then crashing indicates a degredation in functionality as those services used to work before we pushed things remotely.

**Comment by CyrusNajmabadi** [comment]
So i'm ok with this.  THough i would like the code to maybe be commented to explain things.

File: src/Features/Core/Portable/AddImport/AbstractAddImportFeatureService.cs (hunk 2 above)
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/AddImport/AbstractAddImportFeatureService.cs]: I would add a comment saying we're being paranoid here.

File: src/Features/Core/Portable/NavigateTo/AbstractNavigateToSearchService.Remote.cs (hunk 3 above)
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/NavigateTo/AbstractNavigateToSearchService.Remote.cs]: I would add a comment saying we are being paranoid here.

File: src/Workspaces/Core/Portable/FindSymbols/Declarations/DeclarationFinder_AllDeclarations.cs [hunk 4]
```diff
@@ -93,19 +93,22 @@ internal static partial class DeclarationFinder
         private static async Task<(bool, ImmutableArray<SymbolAndProjectId>)> TryFindAllDeclarationsWithNormalQueryInRemoteProcessAsync(
             Project project, SearchQuery query, SymbolFilter criteria, CancellationToken cancellationToken)
         {
-            using (var session = await SymbolFinder.TryGetRemoteSessionAsync(
-                project.Solution, cancellationToken).ConfigureAwait(false))
+            if (RemoteSupportedLanguages.IsSupported(project.Language))
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/FindSymbols/Declarations/DeclarationFinder_AllDeclarations.cs]: this check seems potentially appropriate.  we should comment that this can be called through something like SolutionExplorer and that if this is not C#/VB that the search really can't happen at all (either on the host or remote side).

File: src/Workspaces/Remote/ServiceHub/Services/CodeAnalysisService_DocumentHighlights.cs [hunk 5]
```diff
@@ -16,7 +16,8 @@ internal partial class CodeAnalysisService : IRemoteDocumentHighlights
...
+                documentIdsToSearch.Select(solution.GetDocument).Where(d => d != null));
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Remote/ServiceHub/Services/CodeAnalysisService_DocumentHighlights.cs]: Use .WhereNotNull();

File: src/Workspaces/Remote/ServiceHub/Services/CodeAnalysisService_SymbolFinder.cs [hunk 6]
```diff
@@ -31,6 +31,7 @@ internal partial class CodeAnalysisService : IRemoteSymbolFinder
                 }
 
                 var documents = documentArgs?.Select(solution.GetDocument)
+                                             .Where(d => d != null)
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Remote/ServiceHub/Services/CodeAnalysisService_SymbolFinder.cs]: WhereNotNull

File: src/Workspaces/Remote/ServiceHub/Services/CodeAnalysisService_DocumentHighlights.cs (hunk 1 above)
Comment (CyrusNajmabadi) [inline src/Workspaces/Remote/ServiceHub/Services/CodeAnalysisService_DocumentHighlights.cs]: Add comment explaining the scenario here.

File: src/Workspaces/Remote/ServiceHub/Services/CodeAnalysisService_SymbolFinder.cs (hunk 6 above)
Comment (Pilchie) [inline src/Workspaces/Remote/ServiceHub/Services/CodeAnalysisService_SymbolFinder.cs]: I couldn't find `WhereNotNull`, but maybe it was due to the diagnostic issue Jason was looking at.  I'll try again.

Comment (Pilchie) [inline src/Workspaces/Remote/ServiceHub/Services/CodeAnalysisService_SymbolFinder.cs]: #resolved

File: src/Workspaces/Remote/ServiceHub/Services/CodeAnalysisService_DocumentHighlights.cs (hunk 1 above)
Comment (Pilchie) [inline src/Workspaces/Remote/ServiceHub/Services/CodeAnalysisService_DocumentHighlights.cs]: #resolved

File: src/Workspaces/Core/Portable/FindSymbols/Declarations/DeclarationFinder_AllDeclarations.cs (hunk 4 above)
Comment (Pilchie) [inline src/Workspaces/Core/Portable/FindSymbols/Declarations/DeclarationFinder_AllDeclarations.cs]: Yuck - it only works in the local process because `Project.ContainsSymbolsWithNameAsync` return false for TS projects before we happen to get a `Compilation`.  We should probably move these checks higher up, and make it clearer that they don't work for TS/F#/Xaml in the API docs.

Comment (Pilchie) [inline src/Workspaces/Core/Portable/FindSymbols/Declarations/DeclarationFinder_AllDeclarations.cs]: Filed https://github.com/dotnet/roslyn/issues/20573 to follow up on this more.

**Comment by Pilchie** [comment]
retest windows_debug_vs-integration_prtest please

**Comment by Pilchie** [comment]
retest windows_release_vs-integration_prtest please

**Comment by Pilchie** [comment]
Both failed with Vsixexpinstaller error.

**Comment by Pilchie** [comment]
Tagging @MattGertz for approval.
